packages/expo-notifications/android/src/main/java/expo/modules/notifications/notifications/presentation/builders/ExpoNotificationBuilder.java

```diff
@@ -0,0 +1,157 @@
+
+import android.app.Notification;
+import android.app.NotificationChannel;
+import android.content.Context;
+import android.os.Build;
+import android.os.Bundle;
+import android.provider.Settings;
+import android.util.Log;
+
+import org.json.JSONObject;
+import org.unimodules.core.ModuleRegistry;
+import org.unimodules.interfaces.imageloader.ImageLoader;
+
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import androidx.core.app.NotificationCompat;
+import expo.modules.notifications.notifications.interfaces.NotificationBuilder;
+
+/**
+ * {@link NotificationBuilder} interpreting a JSON request object.
+ */
+public class ExpoNotificationBuilder implements NotificationBuilder {
+  private static final String CONTENT_TITLE_KEY = "title";
+  private static final String CONTENT_TEXT_KEY = "message";
+  private static final String CONTENT_SUBTITLE_KEY = "subtitle";
+  private static final String SOUND_KEY = "sound";
+  private static final String BADGE_KEY = "badge";
+  private static final String BODY_KEY = "body";
+  private static final String THUMBNAIL_URI_KEY = "thumbnailUri";
+
+  private static final String EXTRAS_BADGE_KEY = "badge";
+  private static final String EXTRAS_BODY_KEY = "body";
+
+  private static final long[] NO_VIBRATE_PATTERN = new long[]{0, 0};
+
+  private final Context mContext;
+
+  private JSONObject mNotificationRequest;
+
+  private ImageLoader mImageLoader;
+
+  public ExpoNotificationBuilder(Context context, ModuleRegistry moduleRegistry) {
+    mContext = context;
+    mImageLoader = moduleRegistry.getModule(ImageLoader.class);
+  }
+
+  public ExpoNotificationBuilder setNotificationRequest(JSONObject notificationRequest) {
+    mNotificationRequest = notificationRequest;
+    return this;
+  }
+
+  protected NotificationCompat.Builder createBuilder() {
+    NotificationCompat.Builder builder = new NotificationCompat.Builder(mContext, getChannelId());
+    builder.setSmallIcon(mContext.getApplicationInfo().icon);
+
+    // We're setting the content only if there is anything to set
+    // otherwise the content title and text are displayed
+    // as if they were empty strings.
+    if (!mNotificationRequest.isNull(CONTENT_TITLE_KEY)) {
+      builder.setContentTitle(mNotificationRequest.optString(CONTENT_TITLE_KEY));
+    }
+    if (!mNotificationRequest.isNull(CONTENT_TEXT_KEY)) {
+      builder.setContentText(mNotificationRequest.optString(CONTENT_TEXT_KEY));
+    }
+    if (!mNotificationRequest.isNull(CONTENT_SUBTITLE_KEY)) {
+      builder.setSubText(mNotificationRequest.optString(CONTENT_SUBTITLE_KEY));
+    }
+
+    if (shouldShowAlert()) {
+      // Display as a heads-up notification
+      builder.setPriority(NotificationCompat.PRIORITY_HIGH);
+    } else {
+      // Do not display as a heads-up notification, but show in the notification tray
+      builder.setPriority(NotificationCompat.PRIORITY_DEFAULT);
+    }
```

Comment:
thinking about this some more... I'm leery of this `if (shouldShowAlert())` because I think we want to lean towards giving developers full, explicit control of a platform's capabilities rather than trying to guess at their intent and massage settings implicitly. For an Android developer, it is the case that notifications without an explicit `HIGH` or `MAX` priority flag are not shown as heads-up notifications, whether or not the notification has content title or text. I think in this case it is more confusing for Expo to "paper over" this behavior than to simply expose the underlying platform behavior by requiring that Expo developers, too, set a priority of `high` or `max` if they want a heads-up notification. It's not clear to me that tying a notification's text content (or lack thereof) to its alert behavior is better than just controlling those properties separately, and tying them together may very well lead to confusion.

All of which is to say, I am in favor of removing this `if` block and simply letting developers set a non-default priority explicitly should they choose to. Let me know what you think!

Reply:
I spent a good couple of minutes wondering why would we want to remove the `shouldShowAlert` in favor of plain `priority` only to find out that `shouldShowAlert` implementation isn't what I thought it is — I mistakenly assumed this PR is up-to-date with the _Handling notifications_ PR which introduces the notion of `NotificationBehavior` through which a running JS instance can say _do not show an alert for this notification_. 😃 Now that I think about it I too would be confused about the automatically set priority depending on the presence/absence of text content.

Looking at the API from perspective, what I want to achieve at the end is a way for developers to:

- easily define that the notification which has been received should be displayed as an alert (or not)
- even if the payload sets the priority to `high`
- _after your suggestion_ but also override the priority from payload in JS, if need be

In order to do this the notification will be sent to JS, which will respond with the appropriate behavior along the lines of:
```ts
{
  shouldShowAlert: boolean;
  shouldPlaySound: boolean;
  shouldSetBadge: boolean;
  android?: {
    priority?: number;
  }
}
```
This way the order of significance would be:

> JS.priority > JS.shouldShowAlert > payload.priority

which gives granular control over the priority (`JS.priority`) while also providing an easy, cross-platform interface (`JS.shouldShowAlert`) and a customizable fallback (`payload.priority`).

I will rebase the PR onto `master` since _handling notifications_ already landed there and we can get back to discussing this method.

Reply:
Cool, I think that sounds like a good mix of simple defaults with the ability to be more granular if needed. Thanks for your care in this!!